File: app/src/main/java/com/minerdev/greformanager/HouseRepository.java
package com.minerdev.greformanager;

import android.app.Application;

import androidx.lifecycle.LiveData;

import java.sql.Timestamp;
import java.util.List;

public class HouseRepository {
    private HouseDao houseDao;
    private LiveData<List<House>> allHouses;

    public HouseRepository(Application application) {
        GreDatabase db = GreDatabase.getDatabase(application);
        houseDao = db.houseDao();
        allHouses = houseDao.getAll();
    }

    public LiveData<Timestamp> getLatestUpdatedAt() {
        return houseDao.getLatestUpdatedAt();
    }

    public LiveData<List<House>> getAll() {
        return allHouses;
    }

    public House get(int id) {
        return houseDao.get(id);
    }

    public void insert(House house) {
        GreDatabase.databaseWriteExecutor.execute(() -> houseDao.insert(house));
    }

    public void insert(House... houses) {
        GreDatabase.databaseWriteExecutor.execute(() -> houseDao.insert(houses));
    }

    public void update(House house) {
        GreDatabase.databaseWriteExecutor.execute(() -> houseDao.update(house));
    }

    public void deleteAll() {
        GreDatabase.databaseWriteExecutor.execute(() -> houseDao.deleteAll());
    }
}

File: app/src/main/java/com/minerdev/greformanager/HouseRepository.kt
package com.minerdev.greformanager

import android.app.Application
import androidx.lifecycle.LiveData

class HouseRepository(application: Application) {
    private val houseDao: HouseDao
    val allSale: LiveData<MutableList<House>>
    val allSold: LiveData<MutableList<House>>

    val lastUpdatedAt: Long?
        get() = houseDao.lastUpdatedAt

    operator fun get(id: Int): House? {
        return houseDao[id]
    }

    fun insert(house: House) {
        GreDatabase.databaseWriteExecutor.execute { houseDao.insert(house) }
    }

    fun insert(houses: List<House>) {
        GreDatabase.databaseWriteExecutor.execute { houseDao.insert(houses) }
    }

    fun update(house: House) {
        GreDatabase.databaseWriteExecutor.execute { houseDao.update(house) }
    }

    fun updateOrInsert(house: House) {
        GreDatabase.run {
            databaseWriteExecutor.execute {
                val result = houseDao[house.id]
                if (result == null) {
                    houseDao.insert(house)

                } else {
                    houseDao.update(house)
                }
            }
        }
    }

    fun deleteAll() {
        GreDatabase.databaseWriteExecutor.execute { houseDao.deleteAll() }
    }

    init {
        val db: GreDatabase? = GreDatabase.getDatabase(application)
        houseDao = db!!.houseDao()
        allSale = houseDao.allSale
        allSold = houseDao.allSold
    }
}